### ISSUE TITLE
MD01/heading-increment does not take front matter titles into account as level 1 heading

### ISSUE THREAD

**Issue by afeblot** [DavidAnson/markdownlint#1613]
With markdownlint 0.38 (markdownlint-cli 0.45.0, ) all rules enabled with the default config:

This document generates the expected error:

```md
# level 1

### level 3
```

**mdtest.md:3 MD001/heading-increment Heading levels should only increment by one level at a time [Expected: h2; Actual: h3]**

However, this one where the level 1 heading is replaced by the front-matter title does not:

```md
---
title: "level 1"
---

### level 3
```

Given **MD025/single-title** and **MD041/first-line-heading** do consider this front matter as a level 1 heading, I would expect MD01 to behave consistently and do so as well.

If it does not, I have no clean way the check that pages with a front matter title start with a level 2 heading.

**Comment by DavidAnson** [comment]
MD001 is not appropriate for ensuring the first heading is an H2 whether or not YAML is present. For example, it does not trigger here: https://dlaa.me/markdownlint/#%25m%23%23%20Heading%0A

This test file should show how to achieve what you want using MD041: https://github.com/DavidAnson/markdownlint/blob/224987d727ebada36752074d2466a061c9e03ff0/test/md041-yaml-title-and-first-line-heading.md?plain=1

**Comment by afeblot** [comment]
@DavidAnson , thanks for your quick response.

But it does not help: I confirm your example works;

With this config,

```yaml
first-line-heading:
  level: 2
```

this doc

```md
---
path: "/post"
date: "2012-06-21T10:14:00.000+02:00"
title: "First level heading"
---

## level 2
```

does not trigger a warning. Which is what I expect.

However,

```md
---
path: "/post"
date: "2012-06-21T10:14:00.000+02:00"
title: "First level heading"
---

### level 3
```

does not trigger a warning either. Whereas I would expect a warning here.

What I would like is that:
- the title is considered as the first heading in the page, and to be a level 1 heading (this is the default behavior with MD0025)
- Obviously, be allowed to have multiple level 2 headings.
- A warning if the next heading after the title is not a level 2 (don't we agree this should be reported by default by mardownlint?).

I'm able to achieve this with the below config:

```yaml
default: true

first-line-heading:
  level: 2
  front_matter_title: dontFindTitle
  allow_preamble: true

single-title:
  level: 1
  front_matter_title: dontFindTitle

no-duplicate-heading:
  siblings_only: true
```

This config does trigger a warning on

```md
---
title: "TEST-A"
---

### level 3
```

and as well on

```md
---
title: "TEST-B"
---

# level 1
```

But the triggered rule for these 2 files is `MD041/first-line-heading`. Whereas IMHO, it would be more appropriate to have `MD001/heading-increment` reporting TEST-A, and `MD025/single-title` reporting TEST-B.

And also, obviously, obtaining this result by disabling the front matter detection seems more like a workaround, considering that markdownlint default behavior is to take front matter into account as first heading provider. I assume it is as well reasonnable to consider as a default behavior that after such a title, the next heading level must only be 2, which I why I made my initial report regarding MD001 as a bug.

Where is it that we disagree on these statements, so that the default behavior does not report TEST-B as erroneous?


Thanks

**Comment by DavidAnson** [comment]
Ah, thanks, I hadn't realized you showed the same configuration. The issue with using MD001 here is that its behavior looks to see if a heading level that increases vs. the previous increases by exactly 1. It doesn't currently have an explicit awareness of the first heading in a file. In code, it works by assuming the implied 0th heading level is very large. It could be modified to do what you want by allowing the user to configure the implicit level of the implied 0th heading. But doing so would veer away from the defined purpose of this rule: "This rule is triggered when you skip heading levels in a Markdown document". It probably wouldn't be the worst thing ever to add this and bend the definition slightly, especially as it would be optional.

**Comment by DavidAnson** [comment]
Fixed by #1617
